Made page titles consistent

dwvanstone opened the pull request:
Fixes #1806. Fixes #1125. Fixes #1684.

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/21/

For #1806:

All page titles have the subpage (or the identifying part of the subpage) in the front of the title. For instance, "Preferences - Button Men Online", For specific buttons, profiles, forum threads, etc.,  the specific name will be in the page title. (e.g., a button page: "Spite - Button Men Online"; a profile page "boilerbrown - Button Men Online")

Note that sometimes the HTML page was changed to have the simple title of "Button Men Online". (Such as buttons.html) In these cases the JS is expected to change the title. I found it less visually jarring to have the title change from "Button Men Online" to "Spite - Button Men Online" than from "Buttons - Button Men Online" to "Spite - Button Men Online". 

The overview page title now has the number of active games as part of the page title. ("(2) Button Men Online") The number in parentheses is set to be the sum of open games waiting your move plus the number of games waiting to be accepted by you. It does not include games that have finished and are waiting to be dismissed or games that you're waiting for others to accept. I think this matches the games that the "Redirect to waiting games when in Monitor mode" option would take you to.

For #1125:

The favicon will change to favicon_monitor.ico if the user is in monitor mode on the overview page. (The favicon_monitor.ico represents "I'm in monitor mode".) If the user has pinned the page (where only the favicon shows), they will notice the favicon change only if they have the redirect option on in the monitor mode.

If the redirect option is on in the monitor mode, the page title will be visibly different -- it will either go to the game page ("Game # ...") or to a forum thread page ("UBFC 64 - ..."). If the redirect option is off, the user will observe a difference in the title of the overview page when a game is ready to be played. ("(2) Button Men Online")

For #1684:

The page title now has the forum thread name. This won't help the user who has pinned the page (since only the favicon will appear). But this is at least an improvement for everyone else!

AdmiralJota commented:
Making the Buttons page different from everything else seems odd. I'd advise against it.

What would you think of the idea of making some of these multiple levels deep? E.g., keep the buttons page as "Buttons - Button Men Online", and make an individual button or set into "Spite - Buttons - Button Men Online" or "Soldiers - Buttons - Button Men Online". Similarly for the Forum ("Secret Santa 2015 - Forum - Button Men Online" or "Features and Bugs - Forum - Button Men Online").

Also, you probably also want to modify Login.js, where it sets the dev favicon. I expect we probably only want to have one favicon element in the header at a time.

AdmiralJota commented:
Also, if you back up the old value of the favicon when the Overview page loads into some variable (after Login.js had a chance to potentially set it to dev), then when you're setting the page title based on the number of games waiting for action, you could check if that number > 0, and if so, restore the backed up favicon. Then that will work for people with the tab pinned but without redirecting turned on.

dwvanstone commented:
> Making the Buttons page different from everything else seems odd. I'd advise against it.

What I have currently checked in has the Buttons & Forum the same. As you step into the "hierarchy" only the leaf name shows along with the website name:

    Buttons - Button Men Online
    Dork Victory - Button Men Online
    Carson - Button Men Oline

versus

    Buttons - Button Men Online
    Dork Victory - Buttons - Button Men Online
    Carson - Dork Victory - Buttons -  Button Men Oline

I did the same with Forums:

    Forum - Button Men Online
    Tournaments - Button Men Online
    UBFC heavyweight draft - Button Men Online

versus

    Forum - Button Men Online
    Tournaments - Forum  -Button Men Online
    UBFC heavyweight draft - Tournaments - Forum  - Button Men Online

The user may not have followed the hierarchy to get into the specific button or specific forum thread, so I didn't see what the point was with presenting the hierarchy in the page title. However, I'm happy to be convinced otherwise. Your compromise of at least showing the category ("Spite - Buttons - Button Men Online") is pretty good. 

I'd like you to see it on the dev site before to see how it feels and flows.

> Also, you probably also want to modify Login.js, where it sets the dev favicon

I certainly didn't test what the dev site would look like. To change that I'd need to change all of the .HTML files so let's see how it looks on the dev site. (It's only dev!)

> Also, if you back up the old value of the favicon when the Overview page loads ... Then that will work for people with the tab pinned but without redirecting turned on.

I was going for simplicity of design (the icon simply means "the monitor is on") over meeting that requirement. Can we try it and see how it works on the dev site? Does it feel elegant? 

Thanks for the feedback - this is my first multiple file change!

dwvanstone commented:
I've submitted a fix for the dev site (which can't be tested on Jenkins) to make sure the favicon shows correctly.

I tested on my home branch by modifying the JS code temporarily to pretend it was the DEV branch.

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/22/

AdmiralJota commented:
I think that I might have been thinking of just adding an else clause to Login.js, so that it added the <link> element either way (dev version or prod version), but this works too.

By the way, you can test the site locally as dev. You just need to change Config.js so it says "development" instead of "production".

blackshadowshade commented:
@dwvanstone, I've fixed the opening line so that the Fixes keyword directly prefaces each issue number.

https://help.github.com/articles/closing-issues-via-commit-messages/

dwvanstone commented:
> I've fixed the opening line so that the Fixes keyword directly prefaces each issue number.

Ah, good to know.

cgolubi1 commented:
I owe an apology for not getting this looked at for the December release.

I'll look at it during the merge freeze period, so we can get it on the dev site in December.

cgolubi1 commented on src/ui/buttons.html:
```diff
@@ -6,9 +6,10 @@
 <head>
     <meta charset="utf-8" />
-    <title>Button Men Online &mdash; Buttons</title>
+    <title>Button Men Online</title>
```
Hmm, this looks wrong at a glance - shouldn't it be "Buttons &mdash; Button Men Online"?

dwvanstone replied:
Whatever shows up in buttons.html will be replaced by the JavaScript (unless an error occurs reading from the API). I found having the title flash from "Button Men Online" to "something -- Button Men Online" less jarring than flashing from "Buttons -- Button Men Online" to "something -- Button Men Online". So this was intentional, but I'm happy to change it. (Try it first!)

cgolubi1 commented on src/ui/forum.html:
```diff
@@ -7,8 +7,9 @@
 
     <meta charset="utf-8" />
-    <title>Button Men Online &mdash; Forum</title>
+    <title>Button Men Online</title>
```
Ah, this is the same, so i bet you did it on purpose and are going to generate these headers dynamically - i'll keep reading.  :>)

dwvanstone replied:
Indeed! And I thought having some header was better than no header in case there was an error reaching the server.

cgolubi1 replied:
Yep, you're absolutely right, and you actually said all this in your description - don't stress too much about my intermediate comments while doing a first code read, a lot of them are just notes to myself.

cgolubi1 commented on src/ui/js/Login.js:
```diff
@@ -34,12 +34,7 @@ Login.getLoginHeader = function() {
-            'rel': 'shortcut icon',
-            'href': '/dev_favicon.ico',
-          }));
+        $('#favicon').attr('href','/dev_favicon.ico');
```
Nit: stylistically, i prefer horizontal whitespace between list items and string components in code, so `'href', '/dev_favicon.ico'` rather than `'href','/dev_favicon'`.

(A nit is something that i'm not going to block the pull over, but would like fixed if you have to make other changes and/or find it easy to do so.)

cgolubi1 commented on src/ui/js/Overview.js:
```diff
@@ -94,6 +94,14 @@ Overview.getOverview = function(callback) {
+  if (numGames > 0) {
+    gameCountText='('+numGames+') ';
+  }
+  $('title').html(gameCountText+'Button Men Online');
```
Ditto: this entire block has no horizontal whitespace in variable definitions, or on other side of `+` in string assembly, and i'd like to see it in all those places to match the existing code.

cgolubi1 commented on src/ui/js/Profile.js:
```diff
@@ -21,6 +21,8 @@ Profile.showLoggedInPage = function() {
 Profile.getProfile = function(callback) {
   var playerName = Env.getParameterByName('player');
 
+  $('title').html(playerName+ ' &mdash; Button Men Online');
```
Nit: whitespace

cgolubi1 commented on src/ui/js/Overview.js:
```diff
@@ -134,6 +142,9 @@ Overview.showPage = function() {
 Overview.executeMonitor = function() {
   Api.automatedApiCall = true;
+
+  $('#favicon').attr('href','/favicon_monitor.ico');
```
Nit: whitespace

cgolubi1 commented on test/src/ui/js/test_Overview.js:
```diff
@@ -47,7 +47,8 @@ module("Overview", {
     // Page elements
     $('#overview_page').remove();
-
+    $("#favicon").attr("href","/favicon.ico");
```
Nit: whitespace

cgolubi1 commented:
Test plan:
* Bring up every page type on the site and see if they look visually okay to me, including:
    * Browsing around in the button tree
    * Browsing around in the player tree
    * Browsing around in the forum tree
* Try out the monitor/favicon business:
    * On a prod-configured site with redirect off
    * On a prod-configured site with redirect on
    * On a dev-configured site
    * With a new forum post and no new games

cgolubi1 commented:
Also for future reference: because of how the mac Terminal app handles double-clicking, it's slightly easier for me to deal with branch names with underscores, than branch names with hyphens.  If you don't care, use underscores.  (If you do care, no big deal.)

cgolubi1 commented:
When i look at the game page for a new game (one that players are deciding whether to accept) it shows up as "- Round 1".  That seems wrong - IMO it should say "- New" to be consistent with the game state:
![new_game_in_round_1](https://cloud.githubusercontent.com/assets/2898207/11613748/0bfe4c18-9bfa-11e5-898b-058be18da8bc.png)

cgolubi1 commented:
Similarly: a Cancelled game also shows up as "Round 1" and a Completed game shows up as "Round 3" (or however many rounds there were).  I think you should handle those three non-active game states in the header:
* If in Game.GAME_STATE_CHOOSE_JOIN_GAME, display "* New"
* If in Game.GAME_STATE_END_GAME, display "* Completed"
* If in Game.GAME_STATE_REJECTED, display "* Cancelled"
* Otherwise show the round the way you've been doing

cgolubi1 commented on src/ui/js/Game.js:
```diff
@@ -1664,7 +1664,7 @@ Game.pageAddGameHeader = function(action_desc) {
       Api.game.opponent.button.name + ') ' + Game.SPACE_BULLET +
       'Round #' + Api.game.roundNumber;
-  $('title').html('Button Men Online &mdash; ' + gameTitle);
+  $('title').html(gameTitle + '&mdash; Button Men Online');
```
This is whitespace but not a nit - you need whitespace inside the quotes before the `&mdash;` here, because as is it shows up as `Round #1- Button Men Online` for the player.

cgolubi1 commented:
Okay, i'm out of testing time this morning, so i'll get back to my test plan tomorrow.  Summary of what i've found so far:
* Game page long header should display a different title part instead of "Round #N" for New and Cancelled games (and IMO Completed games too)
* Game page long header needs whitespace before the mdash after the round number
* Code should have horizontal whitespace around `=`, `+`, and `,` in several places for readability

If you have time today and want to fix these now, go for it, and i'll update before continuing.  If you want to wait for the rest of the testing, that's fine too.

dwvanstone commented:
I won't have any time in the next few days to work on this -- I'll wait for all comments and edit everything together.

cgolubi1 commented:
For the record, i feel conflicted about the active game count --- in my tab bar, it looks perfect, but at the top of the browser, it looks weird to me to have this context-free number at the beginning of the bar.  It's possible that i'm just not used to it, and i don't think it's a blocker, just throwing it out there:
![overview_with_active_games](https://cloud.githubusercontent.com/assets/2898207/11626947/de432012-9cb4-11e5-8165-9d1bb214a73f.png)

cgolubi1 commented:
More browsing around:
* If i click on Monitor when i already have active games, i get the inactive icon color
* Something isn't right with active games count - let me reproduce it cleanly in a separate comment

cgolubi1 commented:
Okay, here's what i did:
* Load virtualbox with a database dump i took from the prod site on 20151130
* Login as blackshadowshade

It looks like this - no active games, but the number is showing me 1:
![overview_no_active_games](https://cloud.githubusercontent.com/assets/2898207/11627229/f3f861ae-9cb6-11e5-97bf-afb576741ef6.png)

I'll try to help root-cause that in a bit if i have time, but for now i'll keep going and just note the bug.

cgolubi1 commented:
Okay, in firefox, the favicon monitor mode thing does not work for me.  I consistently see the greyed-out favicon:
* when browsing to monitor mode when i have no active games
* when browsing to monitor mode when i do have active games
* when browsing to monitor mode when i have no active games, advancing a game so i do have an active game, and waiting for the next monitor reload

Did you do any testing where you saw the favicon change in monitor mode?  If so, tell me how you tested it so we can see what's different.

cgolubi1 commented:
Create game page - title looks fine

cgolubi1 commented on src/ui/js/Buttons.js:
```diff
@@ -55,6 +55,8 @@ Buttons.showButton = function() {
     break;
   }
 
+  $('title').html(buttonName + '&mdash; Button Men Online');
```
Here's another whitespace glitch - you need a space between the button name and the mdash

cgolubi1 commented on src/ui/js/Buttons.js:
```diff
@@ -176,6 +178,8 @@ Buttons.showSet = function() {
     break;
   }
 
+  $('title').html(setName + '&mdash; Button Men Online');
```
Ditto - need a space between the setname and the mdash

cgolubi1 commented:
I think the page title should tell you whether you are looking at a player or a button, since players have fanatics.  e.g. "Jota - Button Men Online" is the title whether you're looking at Jota the player or Jota the button (modulo the whitespace bug in the button case).  What about "Jota (player) - Button Men Online"?

I'm willing to merge this as is, just throwing it out in case you have an idea about how to clarify it, or like my idea.

cgolubi1 commented:
I also wonder if the Forum subpages should say in the title that you're in the forum, e.g. "Tournaments (Forum) - Button Men Online".  Again, don't feel strongly.

Anyway, clicking around in the forum seems to work well.

cgolubi1 commented:
Okay, i'm done with the first round of testing.  Here's a summary of all issues in one place (but not sorted by priority):
* Game page long header should display a different title part instead of "Round #N" for New and Cancelled games (and IMO Completed games too, but it's less critical there since round number is meaningful in that case, it's just not the most helpful piece of information)
* Player-visible whitespace is missing before the mdash in the following cases (and look for ones i missed while you're at it):
    * Game page long header after the round number
    * Button page after button name
    * Button page after set name
* Code should have horizontal whitespace around =, +, and , in several places for readability
* Context-free active game count looks strange in the page top header (e.g. "(26) Button Men Online") (willing to merge as-is)
* Active games count is wrong using the prod site database - there's an overcount somewhere
* In firefox (the only browser i've tried), i can't get the favicon-changing behavior in Monitor mode to work under any conditions i've tried
* Subpage headers should tell you which top-level section you are in (and don't) when you're in the following sections (willing to merge as-is):
    * Player
    * Button
    * Forum

cgolubi1 commented:
I also want to mention that this pull demonstrates some of the downsides of a big pull that changes many things and fixes several bugs at once.  Note that there are a bunch of more-or-less nit-level comments above about the exact text contents and whitespace, related to the original issue #1806 you were trying to address with this.

However, the pull also contains the proposed fix for #1125 and the game count business (which i think isn't in response to a particular issue, though maybe it's part of #1125 too, and those may have thorny logic bugs.  We're going to have to fix all three of those types of issues before this can be merged, whereas if it were just a pull for #1806, it could get merged faster.

Not a complaint --- this pull just happens to be a very clear demonstration of how big pulls get blocked in the iteration cycle, and big pulls get blocked in the iteration cycle *frequently*, so i wanted to spell out the example in case it's of use.

AdmiralJota commented:
One suggestion from earlier in the thread was to have a three-part title whenever you're in a sub-page under one of those sections:

    Buttons - Button Men Online
    Soldiers - Buttons - Button Men Online
    Kublai - Buttons - Button Men Online
    Jota - Players - Button Men Online
    Forum - Button Men Online
    Tournaments - Forum - Button Men Online
    UBFC 70 - Forum - Button Men Online

AdmiralJota commented:
For the Overview game count, maybe format it a little more similarly to the other titles?

    (3 games waiting) - Button Men Online

You'd still see the number first on the tab, but the full title at the top of the page wouldn't be context-free anymore (and, as a bonus, it'd be more consistent with other pages on the site).

dwvanstone commented:
AdmiralJota: I like the three-part titles. I'll go ahead and work on those.
I also like your suggestion for "(3 games waiting)" as well. 

> In firefox (the only browser i've tried), i can't get the favicon-changing behavior in Monitor mode to work under any conditions i've tried

Right now the Monitor mode favicon is coded so the favicon is in black & white if the user is in Monitor mode -- nothing to do with whether there are games waiting, so I think all of your tests are consistent with the code. The idea was that the favicon means "I'm in monitor mode". It sounds like you were expecting the favicon to be black and white to mean "I'm in monitor mode but no games are waiting for me". I can look into that change as well.

> Active games count is wrong using the prod site database - there's an overcount somewhere

I may need help in sorting this one out, but we can talk about that later.

I'll try to make a pass at this tomorrow morning.

I see your point about the downsides of a large pull, but I was having a hard time separating out the bug fixes without the grand sweep of changing all of the titles at once. For instance, I couldn't only fix the Forum titles without having the change be consistent with the way other titles are changed...and I'm fine if those fixes wait until the overall titles are changed. Also, the monitor mode changes involved both the favicon changes and the title changes, but I could have put those into a separate pull. I may go ahead and separate these out if it's easier for you for testing.

irilyth commented:
> Right now the Monitor mode favicon is coded so the favicon is in black & white if the user is in Monitor mode -- nothing to do with whether there are games waiting, so I think all of your tests are consistent with the code. The idea was that the favicon simply means "I'm in monitor mode". It sounds like you were expecting the favicon to be black and white to mean "I'm in monitor mode but no games are waiting for me".

That'd be great -- that's what I had in mind in #1125, I think I didn't realize that "in monitor mode" meant something other than "looking at the overview page and it says 'Monitor Active', which only happens if you don't have any active games", sorry if "in monitor mode" was a confusing way to phrase that.

dwvanstone commented:
> That'd be great -- that's what I had in mind in #1125, I think I didn't realize that "in monitor mode" meant something other than "looking at the overview page and it says 'Monitor Active', which only happens if you don't have any active games", sorry if "in monitor mode" was a confusing way to phrase that.

I was trying to be consistent to when the Overview page said "* Monitor Active* " which is says whether or not there are active games.

irilyth commented:
Hmm. What's the scenario when the overview page says "Monitor Active", and you *do* have active games? Wouldn't it take you to your first active game in that case?

irilyth commented:
Oh, maybe not, if you for some reason have "Redirect to waiting games when in Monitor mode" unchecked. I don't have any idea why anyone uses that preference, so I had forgotten about it. :^)

dwvanstone commented:
> Hmm. What's the scenario when the overview page says "Monitor Active", and you do have active games? Wouldn't it take you to your first active game in that case?

Only if you have "Redirect to waiting games when in Monitor mode" set. In which case the favicon changing doesn't matter on the Overview page.

cgolubi1 commented:
On Mon, 07 Dec, 2015 at 07:16:39 -0800, dwvanstone wrote:

> I see your point about the downsides of a large pull, but I was having a hard time separating out the bug fixes without the grand sweep of changing all of the titles at once. For instance, I couldn't only fix the Forum titles without having the change be consistent with the way other titles are changed...and I'm fine if those fixes wait until the overall titles are changed. Also, the monitor mode changes involved both the favicon changes and the title changes, but I could have put those into a separate pull. I may go ahead and separate these out if it's easier for you for testing.

Nah.  Now that i've looked this over, i can keep all the threads straight in my head (and my notes), so no point splitting it unless you want to for your own purposes.  I agree that it's a difficult dance.

At this point, i *would* suggest reverting the monitor to just using the normal favicon, and if we want to modify that favicon based on what's going on, we do so in a different pull.

And, yeah, i'll try to help give a real example for the count bug.  It's a blocker, but it'll probably be pretty time-consuming for you to find the bug with the minimal info i've given you, and without access to the prod DB.

dwvanstone commented:
> At this point, i *would* suggest reverting the monitor to just using the normal favicon, and if we want to modify that favicon based on what's going on, we do so in a different pull.

Just to check, you're suggesting I remove the code that updates the favicon for a different pull? Or are you suggesting that I code it so the favicon switches from black & white to normal if there are games waiting? (The first is simpler, of course!)

cgolubi1 commented:
On Mon, 07 Dec, 2015 at 07:32:27 -0800, dwvanstone wrote:

> Just to check, you're suggesting I remove the code that updates the favicon for a different pull? Or are you suggesting that I code it so the favicon switches from black & white to normal if there are games waiting? (The first is simpler, of course!)

I can't parse this sentence, but i think i'm suggesting the former - make the monitor page go back to using favicon.ico (or dev_favicon.ico) rather than favicon_monitor.ico.

I'm not looking at the code right now, but IIRC your pull includes some plumbing to make it possible to vary the favicon within code --- i'm happy with that plumbing being left in the pull for later use --- it appears to work.

dwvanstone commented:
Got it! (I will try to make more parseable sentences in the future...!)

cgolubi1 commented on src/ui/js/Overview.js:
```diff
@@ -94,6 +94,14 @@ Overview.getOverview = function(callback) {
 
+  var gameCountText;
+  var numGames=Api.new_games.nGames;
+  numGames+=Api.active_games.games.awaitingPlayer.length;
```
Oh, i see the problem here - `Api.new_games` contains both new games which are awaiting your action *and* those which are awaiting the other player's action.  It's just like `Api.active_games` that way, and, as in that case, `Api.new_games.games.awaitingPlayer.length` is defined.  (However, for either array, i think you get undefined rather than 0 when there are no games, so you should probably not count on lazy casting to do the right thing, but instead define numGames as 0 and check whether each array is defined before adding its length, or something along those lines.)

dwvanstone commented:
> Active games count is wrong using the prod site database - there's an overcount somewhere

I wonder if in the prod site database there is a stale game which is waiting for action. Otherwise there's something not right coming in from the back end -- I'm simply adding Api.new_games.nGames with Api.active_games.games.awaitingPlayer.length.

cgolubi1 commented:
> I wonder if in the prod site database there is a stale game which is waiting for action. Otherwise there's something not right coming in from the back end -- I'm simply adding Api.new_games.nGames with Api.active_games.games.awaitingPlayer.length.

See the comment i made half an hour ago --- you can't use Api.new_games.nGames because that contains new games which aren't waiting for the player's action.

dwvanstone commented:
Heh - I didn't see that comment yet - we're both up early working on this. Thank you!!

Just to check, did we agree that games are "cancelled" or is there a new adjective/verb/grammar-thing to describe games like this?

cgolubi1 commented:
> Just to check, did we agree that games are "cancelled" or is there a new adjective/verb/grammar-thing to describe games like this? 

We agreed that "cancelled" is the word to describe the state of games which a player either rejected or withdrew before play started.

But, just checking: that's not what Api.new_games is about --- Api.new_games is about games which are waiting for one of the players to decide whether to play them.  And the point is that if you proposed a game and are waiting for your opponent to decide whether to play it, that game is not waiting for action from you, so it shouldn't show up in the count of games waiting for you.

dwvanstone commented:
No worries - I was asking in order to fix the overview header/title for when the game has been rejected.

I have a fix for the number if games awaiting.

dwvanstone commented:
I have made all of the requested fixes.

Note that by updating the page title for the Game page I've also updated the header for the Game page. Thus for new, cancelled, and completed games both the title and the header will have the new text.

I have removed the change for the favicon for now -- we can address this in a different fix.

Thanks for your patience while I learn the ropes!

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/22/

Do I need to merge branches? (I see: "This branch has conflicts that must be resolved")

cgolubi1 commented:
Yes, you need to merge in changes from upstream and resolve conflicts.  This will look something like:
```
git pull upstream master
```
and then read what git tells you.  It'll say you have merge conflicts and you need to resolve them and then do something.  Hang onto the information about what it says you should run after you resolve the conflicts.

To resolve conflicts, look at each file that it tells you has conflicts, and search for `>>>` --- it'll give you some syntax that will hopefully be familiar from other merge conflict syntaxes, showing the code that it doesn't know what to do about.

Then do whatever it told you you had to do to get your working dir into a clean state, then push to your branch on github again.

Feel free to swing by IRC sometime to get help --- merge conflicts are a hassle to deal with until you're used to them, so you'll maximize your own happiness by getting used to them so they aren't a hassle.  :>)

blackshadowshade commented:
That bit about 'what it tells you' is probably the following:
- add the conflicted file explicitly using 'git add <filepath>'
- then do a full commit after the merge

dwvanstone commented:
Hmmm, I'm not happy with how the merging ended up. I didn't quite do it
right, and will have to try again to do it differently in the future.

The merging process didn't work too well -- after staring at the ">>>" and
"<<<" code and not understanding what to do I thought it would be simpler
to simply clone master, copy over the offending files from master to my
branch (Api.js and Overview.js), and make my few lines of changes for my
fix. Everything works fine in my branch

However, it looks like my pull request has tons of changes in Api.js and
Overview.js which is not the case. I thought I was being clever. I may need
to back out and start all over again from master and re-make my changes.  :(

That's a project for another day.

-dvs-


On Tue, Dec 8, 2015 at 8:46 AM, cgolubi1 <notifications@github.com> wrote:

> Yes, you need to merge in changes from upstream and resolve conflicts.
> This will look something like:
>
> git pull upstream master
>
> and then read what git tells you. It'll say you have merge conflicts and
> you need to resolve them and then do something. Hang onto the information
> about what it says you should run after you resolve the conflicts.
>
> To resolve conflicts, look at each file that it tells you has conflicts,
> and search for >>> --- it'll give you some syntax that will hopefully be
> familiar from other merge conflict syntaxes, showing the code that it
> doesn't know what to do about.
>
> Then do whatever it told you you had to do to get your working dir into a
> clean state, then push to your branch on github again.
>
> Feel free to swing by IRC sometime to get help --- merge conflicts are a
> hassle to deal with until you're used to them, so you'll maximize your own
> happiness by getting used to them so they aren't a hassle. :>)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/buttonmen-dev/buttonmen/pull/1821#issuecomment-162884035>
> .
>

cgolubi1 commented:
Okay, so basically what's going on is that your diff shows all the changes from master which you merged in.  This happens to shadowshade all the time, and he has a process for getting github to show a clean diff with only his changes.

Shadowshade, can you describe how you do that?  DVS, i don't think you need to start over --- i think you're close to having this working.

blackshadowshade commented:
The easiest thing to do at the moment is do another 'git pull upstream master' and then 'git push origin HEAD'. This should work because Chaos has merged irilyth's pull request, so master now looks different, and so this forces Github to reset its diff.

See if that works.

irilyth commented:
> The easiest thing to do at the moment is do another 'git pull upstream master' and then 'git push origin HEAD'. This should work because Chaos has merged irilyth's pull request, so master now looks different, and so this forces Github to reset its diff.

I'm very close to understanding what's going on here, but I don't entirely, because my workflow at my day job happens to involve a lot more rebasing than merging. I'd love to take a look at your current state, DVS, if you haven't already lost it. Maybe swing by IRC some time?

(In particular, I have a hypothesis about what 'git push origin HEAD' would do in this case, and why it would help, but I'd need to look at some state to be sure.)

cgolubi1 commented:
So, based on conversation on IRC last night, i think this was updated and the merge conflict fixed, but devious wanted to look at the code some more before having me look at it.  So i'm going to treat it as "still incomplete" for now --- let me know if that's not true/when i should look at it again.

dwvanstone commented:
Thanks to everybody's help I succeeded in wrestling with git, but I've been unable to merge my changes with the new Api.js code and need N*10 minutes of focused attention on it for some value of N. I then need to pass everything through jenkins...we'll see if I can spend that time tonight or tomorrow morning.

dwvanstone commented:
http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/26/

It has passed!

cgolubi1 commented:
Great.  I'll take a look in the morning.

cgolubi1 commented:
What i need to do here:

Verify fixes for all the previously-mentioned issues:
* Game page long header should display a different title part instead of "Round #N" for New and Cancelled games (and IMO Completed games too, but it's less critical there since round number is meaningful in that case, it's just not the most helpful piece of information)
* Player-visible whitespace is missing before the mdash in the following cases (and look for ones i missed while you're at it):
    * Game page long header after the round number
    * Button page after button name
    * Button page after set name
* Code should have horizontal whitespace around =, +, and , in several places for readability
* Context-free active game count looks strange in the page top header (e.g. "(26) Button Men Online") (willing to merge as-is)
* Active games count is wrong using the prod site database - there's an overcount somewhere
* In firefox (the only browser i've tried), i can't get the favicon-changing behavior in Monitor mode to work under any conditions i've tried
* Subpage headers should tell you which top-level section you are in (and don't) when you're in the following sections (willing to merge as-is):
    * Player
    * Button
    * Forum

Look at the new commits and make sure the changes look okay.

cgolubi1 commented:
![screencap](https://cloud.githubusercontent.com/assets/2898207/11743558/78a65db2-9fd6-11e5-8a80-3eedbb1abfa7.png)

cgolubi1 commented:
Per discussion on IRC: devious and i both think we should probably go back to "(23) - Button Men Online" because it looks better in the tab bar.  But i'll merge it as is if i don't find any other issues, and dev testers can weigh in.

dwvanstone commented:
I'm not too happy with the way "(N games waiting) -" appears in my Safari tab bar -- it doesn't strike me as "this is the ButtonMen tab". My preference would be to return it to "(N) Button Men Online" as I had before.

![screen shot 2015-12-11 at 7 22 01 am](https://cloud.githubusercontent.com/assets/6751265/11743764/084d2238-9fd8-11e5-8fc7-5c8f970245c9.png)

cgolubi1 commented:
dwvanstone: yep, agreed.

cgolubi1 commented:
I think that's everything.  Let's try this, and i'll create a new issue for changing the game count header.

AdmiralJota commented:
"(23) - Button Men Online" or "(23) Button Men Online"? I prefer the former, personally.
